crates/ra_ide/src/imports_locator.rs

```diff
@@ -0,0 +1,97 @@
+//! This module contains an import search funcionality that is provided to the ra_assists module.
+//! Later, this should be moved away to a separate crate that is accessible from the ra_assists module.
+
+use crate::{
+    db::RootDatabase,
+    references::{classify_name, classify_name_ref, NameDefinition, NameKind},
+    symbol_index::{self, FileSymbol},
+    Query,
+};
+use ast::NameRef;
+use hir::{db::HirDatabase, InFile, ModPath, Module, SourceBinder};
+use itertools::Itertools;
+use ra_assists::ImportsLocator;
+use ra_prof::profile;
+use ra_syntax::{ast, AstNode, SyntaxKind::NAME};
+
+pub(crate) struct ImportsLocatorIde<'a> {
+    source_binder: SourceBinder<'a, RootDatabase>,
+}
+
+impl<'a> ImportsLocatorIde<'a> {
+    pub(crate) fn new(db: &'a RootDatabase) -> Self {
+        Self { source_binder: SourceBinder::new(db) }
+    }
+
+    fn search_for_imports(
+        &mut self,
+        name_to_import: &ast::NameRef,
+        module_with_name_to_import: Module,
+    ) -> Vec<ModPath> {
+        let _p = profile("search_for_imports");
+        let db = self.source_binder.db;
+        let name_to_import = name_to_import.text();
+
+        let project_results = {
+            let mut query = Query::new(name_to_import.to_string());
+            query.exact();
+            query.limit(10);
+            symbol_index::world_symbols(db, query)
+        };
+        let lib_results = {
+            let mut query = Query::new(name_to_import.to_string());
+            query.libs();
+            query.exact();
+            query.limit(10);
+            symbol_index::world_symbols(db, query)
+        };
+
+        project_results
+            .into_iter()
+            .chain(lib_results.into_iter())
+            .filter_map(|import_candidate| self.get_name_definition(db, &import_candidate))
+            .filter_map(|name_definition_to_import| {
+                if let NameKind::Def(module_def) = name_definition_to_import.kind {
+                    module_with_name_to_import.find_use_path(db, module_def)
```

Comment:
note that the use path may actually end in a different segment than the original name, if it's using a reexport, so we should probably filter that out, or change the actual path in the assist, or maybe later add a parameter to `find_use_path` that disregards such paths

Reply:
I'm afraid I cannot imagine this myself, do you have any code snippet I can try out and fix it?